File: database_integration_scripts/enhance_db/src/bin/15_reorder_sitelinks.rs
/// Reorder sitelinks columns: id, wikidata_id, individual_name, site, title, url
use anyhow::Result;
use rusqlite::Connection;
use std::fs;
use std::io::Write;
use std::time::Instant;

const DB_PATH: &str = "data/humans_clean.sqlite3";
const TASK_LOG: &str = "task.log";

fn log(msg: &str) {
    let now = chrono_now();
    let line = format!("[{}] {}", now, msg);
    println!("{}", line);
    let mut f = fs::OpenOptions::new()
        .create(true)
        .append(true)
        .open(TASK_LOG)
        .unwrap();
    writeln!(f, "{}", line).unwrap();
}

fn chrono_now() -> String {
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap();
    let secs = dur.as_secs();
    let hours = (secs % 86400) / 3600;
    let mins = (secs % 3600) / 60;
    let s = secs % 60;
    format!("{:02}:{:02}:{:02} UTC", hours, mins, s)
}

fn elapsed(start: Instant) -> String {
    let d = start.elapsed();
    let secs = d.as_secs();
    if secs < 60 {
        format!("{}s", secs)
    } else if secs < 3600 {
        format!("{}m {}s", secs / 60, secs % 60)
    } else {
        format!("{}h {}m {}s", secs / 3600, (secs % 3600) / 60, secs % 60)
    }
}

fn main() -> Result<()> {
    let total_start = Instant::now();
    log("=== Step 15: Reorder sitelinks columns ===");

    let conn = Connection::open(DB_PATH)?;
    conn.execute_batch(
        "PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA cache_size=-2000000;",
    )?;

    let total: i64 = conn.query_row("SELECT COUNT(*) FROM sitelinks", [], |r| r.get(0))?;
    log(&format!("[15] Sitelinks table has {} rows", total));

    let mut stmt = conn.prepare("PRAGMA table_info(sitelinks)")?;
    let cols_before: Vec<String> = stmt
        .query_map([], |r| r.get::<_, String>(1))?
        .filter_map(|r| r.ok())
        .collect();
    log(&format!("[15] Current columns: {}", cols_before.join(", ")));

    let step = Instant::now();
    log("[15] Restructuring sitelinks (moving individual_name after wikidata_id)...");

    conn.execute_batch("ALTER TABLE sitelinks RENAME TO sitelinks_backup;")?;
    log(&format!("[15]   Renamed to backup ({})", elapsed(step)));

    conn.execute_batch(
        "CREATE TABLE sitelinks (
             id INTEGER PRIMARY KEY AUTOINCREMENT,
             wikidata_id TEXT NOT NULL,
             individual_name TEXT,
             site TEXT,
             title TEXT,
             url TEXT
         );",
    )?;

    log("[15]   Inserting data...");
    let step = Instant::now();
    conn.execute_batch(
        "INSERT INTO sitelinks (id, wikidata_id, individual_name, site, title, url)
         SELECT id, wikidata_id, individual_name, site, title, url
         FROM sitelinks_backup;",
    )?;
    log(&format!("[15]   INSERT complete ({} for {} rows)", elapsed(step), total));

    let step = Instant::now();
    conn.execute_batch("DROP TABLE sitelinks_backup;")?;
    log(&format!("[15]   Backup dropped ({})", elapsed(step)));

    // Recreate index
    let step = Instant::now();
    conn.execute_batch(
        "CREATE INDEX IF NOT EXISTS idx_sitelinks_wikidata ON sitelinks(wikidata_id);",
    )?;
    log(&format!("[15]   Index created ({})", elapsed(step)));

    let verify: i64 = conn.query_row("SELECT COUNT(*) FROM sitelinks", [], |r| r.get(0))?;
    log(&format!("[15] Sitelinks after restructure: {} rows", verify));

    let mut stmt2 = conn.prepare("PRAGMA table_info(sitelinks)")?;
    let cols_after: Vec<String> = stmt2
        .query_map([], |r| r.get::<_, String>(1))?
        .filter_map(|r| r.ok())
        .collect();
    log(&format!("[15] New columns: {}", cols_after.join(", ")));

    log(&format!(
        "=== Step 15 complete (total: {}) ===",
        elapsed(total_start)
    ));
    Ok(())
}

File: scripts/database_integration_scripts_V2/legacy/15_reorder_sitelinks.py
"""15 - Reorder sitelinks columns: id, wikidata_id, individual_name, ...

Mirrors `enhance_db/src/bin/15_reorder_sitelinks.rs`.

  Inputs : sitelinks (any earlier column order)
  Output : sitelinks rebuilt with column order
             id, wikidata_id, individual_name, site, title, url
           and idx_sitelinks_wikidata recreated.

Usage
-----
    python3 15_reorder_sitelinks.py
    python3 15_reorder_sitelinks.py --full
"""

from __future__ import annotations

import sqlite3
import tempfile
from pathlib import Path

from common import insert_rows, log, open_db, parse_run_mode


def run(conn: sqlite3.Connection) -> None:
    log("=== Step 15: Reorder sitelinks columns ===")

    total = conn.execute("SELECT COUNT(*) FROM sitelinks").fetchone()[0]
    log(f"[15] Sitelinks table has {total} rows")
    cols_before = [r[1] for r in conn.execute("PRAGMA table_info(sitelinks)").fetchall()]
    log(f"[15] Current columns: {', '.join(cols_before)}")

    log("[15] Restructuring sitelinks (moving individual_name after wikidata_id)...")
    conn.executescript(
        """
        DROP TABLE IF EXISTS sitelinks_backup;
        ALTER TABLE sitelinks RENAME TO sitelinks_backup;

        CREATE TABLE sitelinks (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            wikidata_id TEXT NOT NULL,
            individual_name TEXT,
            site TEXT,
            title TEXT,
            url TEXT
        );

        INSERT INTO sitelinks (id, wikidata_id, individual_name, site, title, url)
        SELECT id, wikidata_id, individual_name, site, title, url
        FROM sitelinks_backup;

        DROP TABLE sitelinks_backup;
        """
    )
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_sitelinks_wikidata ON sitelinks(wikidata_id)"
    )
    conn.commit()

    verify = conn.execute("SELECT COUNT(*) FROM sitelinks").fetchone()[0]
    log(f"[15] Sitelinks after restructure: {verify} rows")
    cols_after = [r[1] for r in conn.execute("PRAGMA table_info(sitelinks)").fetchall()]
    log(f"[15] New columns: {', '.join(cols_after)}")
    log("=== Step 15 complete ===")


def _sample_main() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        db_path = Path(tmp) / "sample.sqlite3"
        with sqlite3.connect(db_path) as seed:
            seed.execute(
                "CREATE TABLE sitelinks ("
                "id INTEGER PRIMARY KEY AUTOINCREMENT, "
                "site TEXT, title TEXT, url TEXT, "
                "wikidata_id TEXT NOT NULL, individual_name TEXT)"
            )
            insert_rows(seed, "sitelinks", [
                {"id": 1, "site": "enwiki", "title": "Alice",
                 "url": "https://en.wikipedia.org/wiki/Alice",
                 "wikidata_id": "Q1", "individual_name": "Alice"},
                {"id": 2, "site": "frwiki", "title": "Bob",
                 "url": "https://fr.wikipedia.org/wiki/Bob",
                 "wikidata_id": "Q2", "individual_name": "Bob"},
            ])
            seed.commit()

        with open_db(db_path) as conn:
            run(conn)
            rows = conn.execute(
                "SELECT id, wikidata_id, individual_name, site, title, url FROM sitelinks"
            ).fetchall()
        for r in rows:
            log(f"  sitelinks: {r}")


if __name__ == "__main__":
    if parse_run_mode() == "full":
        with open_db() as conn:
            run(conn)
    else:
        _sample_main()
